gen/builders/actors.go

```diff
@@ -143,86 +164,123 @@ type MinerActorCfg struct {
-	a.bc.Assert.NoError(err, "seal proof sector size")
-
-	ps, err := builtin.SealProofWindowPoStPartitionSectors(cfg.SealProofType)
-	a.bc.Assert.NoError(err, "seal proof window PoSt partition sectors")
-
-	mi := &miner.MinerInfo{
-		Owner:                      owner.ID,
-		Worker:                     worker.ID,
-		PendingWorkerKey:           nil,
-		PeerId:                     abi.PeerID("test"),
-		Multiaddrs:                 nil,
-		SealProofType:              cfg.SealProofType,
-		SectorSize:                 ss,
-		WindowPoStPartitionSectors: ps,
-	}
-	infoCid, err := a.st.Stores.CBORStore.Put(context.Background(), mi)
-	if err != nil {
-		panic(err)
-	}
+	var (
+		owner  = a.Account(address.SECP256K1, cfg.OwnerBalance)
+		worker = a.Account(address.BLS, big.Zero())
 
-	// create the miner actor s.t. it exists in the init actors map
-	minerState, err := miner.ConstructState(infoCid,
-		cfg.PeriodBoundary,
-		EmptyBitfieldCid,
-		EmptyArrayCid,
-		EmptyMapCid,
-		EmptyDeadlinesCid,
-		EmptyVestingFundsCid,
+		minerInfo cbor.Marshaler
+		state     cbor.Marshaler
+		err       error
 	)
-	if err != nil {
-		panic(err)
-	}
 
-	// TODO allow an address to create multiple miners.
-	minerActorAddr := worker.NextActorAddress(0, 0)
-	handle := a.st.CreateActor(builtin.StorageMinerActorCodeID, minerActorAddr, big.Zero(), minerState)
-
-	// assert miner actor has been created, exists in the state tree, and has an entry in the init actor.
-	// next update the storage power actor to track the miner
-
-	var spa power.State
-	a.st.ActorState(builtin.StoragePowerActorAddr, &spa)
-
-	// set the miners claim
-	hm, err := adt.AsMap(adt.WrapStore(context.Background(), a.st.Stores.CBORStore), spa.Claims)
-	if err != nil {
-		panic(err)
+	switch a.st.ActorsVersion {
+	case actors.Version0:
+		minerInfo, err = miner0.ConstructMinerInfo(owner.ID, worker.ID, nil, []byte("test"), nil, cfg.SealProofType)
+	case actors.Version2:
+		minerInfo, err = miner2.ConstructMinerInfo(owner.ID, worker.ID, nil, []byte("test"), nil, cfg.SealProofType)
+	default:
+		panic("unknown actors version")
 	}
 
-	// add claim for the miner
-	// TODO: allow caller to specify.
-	err = hm.Put(abi.AddrKey(handle.ID), &power.Claim{
-		RawBytePower:    abi.NewStoragePower(0),
-		QualityAdjPower: abi.NewStoragePower(0),
-	})
-	if err != nil {
-		panic(err)
-	}
+	a.bc.Assert.NoError(err, "failed to construct miner info")
 
-	// save the claim
-	spa.Claims, err = hm.Root()
+	infoCid, err := a.st.Stores.CBORStore.Put(context.Background(), minerInfo)
 	if err != nil {
 		panic(err)
 	}
 
-	// update miner count
-	spa.MinerCount += 1
-
-	// update storage power actor's state in the tree
-	_, err = a.st.Stores.CBORStore.Put(context.Background(), &spa)
-	if err != nil {
-		panic(err)
+	// TODO allow an address to create multiple miners.
+	minerActorAddr := worker.NextActorAddress(0, 0)
+	var minerHandle AddressHandle
+	switch a.st.ActorsVersion {
+	case actors.Version0:
```

Comment:
Unfortunately, state constructors are not abstracted, so we have to rely in physical actor versions. @Stebalien